naga/src/front/spv/image.rs

```diff
@@ -486,9 +488,34 @@ impl<I: Iterator<Item = u32>> super::Frontend<I> {
                     let lod_lexp = self.lookup_expression.lookup(lod_expr)?;
                     let lod_handle =
                         self.get_expr_handle(lod_expr, lod_lexp, ctx, emitter, block, body_idx);
+
+                    let is_depth_image = {
+                        let image_lexp = self.lookup_sampled_image.lookup(sampled_image_id)?;
+                        let image_ty = ctx.get_image_expr_ty(image_lexp.image)?;
+                        matches!(
+                            ctx.type_arena[image_ty].inner,
+                            crate::TypeInner::Image {
+                                class: crate::ImageClass::Depth { .. },
+                                ..
+                            }
+                        )
+                    };
+
                     level = if options.compare {
                         log::debug!("Assuming {:?} is zero", lod_handle);
                         crate::SampleLevel::Zero
+                    } else if is_depth_image {
+                        log::debug!(
+                            "Assuming level {:?} converts losslessly to an integer",
+                            lod_handle
+                        );
+                        let expr = crate::Expression::As {
+                            expr: lod_handle,
+                            kind: crate::ScalarKind::Sint,
+                            convert: Some(4),
+                        };
```

Comment:
I think that if we `round` it first this will behave the same as the sampler doing it.

Reply:
My assumption was that a fractional lod interpolated between two mip levels. I can't see anything in the spec which says that (but nor that it rounds the lod). Do you happen to know what the spec says?

Reply:
Depth textures can only be paired with a non-filtering (or comparison) sampler which means that the `mipmapFilter` field of the sampler descriptor can't be `linear`, it must be `nearest` which sounds like it `round`s instead.

A few links:

https://www.w3.org/TR/webgpu/#ref-for-dom-gpusamplerbindingtype-filtering%E2%91%A2
https://www.w3.org/TR/webgpu/#dom-gpusamplerbindingtype-non-filtering
https://www.w3.org/TR/webgpu/#ref-for-dom-gpusampler-isfiltering-slot
https://www.w3.org/TR/webgpu/#enumdef-gpumipmapfiltermode

Reply:
So that's a constraint in webgpu, but is the important thing here not what (if any) constraints exist in SPIR-V so we can best match the behaviour of the input SPIR-V shader when converting to WGSL? If SPIRV allows floating point values then we're potentially changing the behaviour of the input shader, regardless of whether we round to nearest or just cast (which rounds towards zero?)

FWIW I think tint simply casts: https://dawn.googlesource.com/dawn/+/refs/heads/main/src/tint/lang/spirv/reader/ast_parser/function.cc#5661

I don't think it makes much difference either way so happy to add a round, but just want to check I'm understanding the problem correctly.

Reply:
> If SPIRV allows floating point values then we're potentially changing the behaviour of the input shader, regardless of whether we round to nearest or just cast (which rounds towards zero?)

That's fair. I found this in the docs of `VK_FORMAT_FEATURE_SAMPLED_IMAGE_FILTER_LINEAR_BIT`:

> If the format being queried is a depth/stencil format, this bit only specifies that the depth aspect (not the stencil aspect) of an image of this format supports linear filtering, and that linear filtering of the depth aspect is supported whether depth compare is enabled in the sampler or not. Where depth comparison is supported it <strong class="purple">may</strong> be linear filtered whether this bit is present or not, but where this bit is not present the filtered value <strong class="purple">may</strong> be computed in an implementation-dependent manner which differs from the normal rules of linear filtering. The resulting value <strong class="purple">must</strong> be in the range <span class="eq">[0,1]</span> and <strong class="purple">should</strong> be proportional to, or a weighted average of, the number of comparison passes or failures.

It sounds like depth textures can support linear filtering with non-comparison samplers in Vulkan. So yes, there doesn't seem to be a way to preserve the same semantics of SPIR-V & Vulkan. This is only a problem if naga is used with something else other than wgpu since it would be a validation error in wgpu.

I'm not sure what the best solution is here, maybe we leave things as they are and if someone runs into this we can consider other solutions then. If you think so too, resolve this conversation and the PR will land (I will approve the PR and enable auto-merge).

Reply:
That sounds like a reasonable approach to me. Thanks!